packages/react-compose/src/types.ts

```diff
@@ -9,17 +14,31 @@ export type PropsOfElement<
   E extends keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>
 > = JSX.LibraryManagedAttributes<E, React.ComponentPropsWithRef<E>>;
 
-export type ComponentWithAs<E extends React.ElementType, P> = (<EE extends React.ElementType = E>(
-  props: Omit<PropsOfElement<EE>, 'as' | keyof P> & { as?: EE } & P,
-) => JSX.Element) & {
+// export type ComponentWithAs<E extends React.ElementType = 'div', P = {}> = (<EE extends React.ElementType = E>(
+//   props: Omit<PropsOfElement<EE>, 'as' | keyof P> & { as?: EE } & P,
+// ) => JSX.Element) & {
+//   displayName?: string;
+//
+//   defaultProps?: Partial<P & { as: E }>;
+//   propTypes?: React.WeakValidationMap<P> & {
+//     // tslint:disable-next-line:no-any
+//     as: React.Requireable<string | ((props: any, context?: any) => any) | (new (props: any, context?: any) => any)>;
+//   };
+// };
+
+// tslint:disable-next-line:interface-name
+export interface ComponentWithAs<E extends React.ElementType = 'div', P = {}> extends React.FunctionComponent {
```

Comment:
@layershifter what do you think about this?

Reply:
If it works on all scenarios it will be great. But as I remember `interface` breaks type safety

Reply:
Would love to know more about when the right time to use type vs interface is.